File: ask-o11y/pkg/rbac/rbac_test.go
package rbac

import (
	"consensys-asko11y-app/pkg/mcp"
	"testing"
)

func boolPtr(b bool) *bool { return &b }

func readOnlyTool(name string) mcp.Tool {
	return mcp.Tool{
		Name:        name,
		Annotations: &mcp.ToolAnnotations{ReadOnlyHint: boolPtr(true)},
	}
}

func writeTool(name string) mcp.Tool {
	return mcp.Tool{
		Name:        name,
		Annotations: &mcp.ToolAnnotations{ReadOnlyHint: boolPtr(false)},
	}
}

func unannotatedTool(name string) mcp.Tool {
	return mcp.Tool{Name: name}
}

func TestIsReadOnlyTool(t *testing.T) {
	tests := []struct {
		name string
		tool mcp.Tool
		want bool
	}{
		{"annotated read-only", readOnlyTool("get_dashboard"), true},
		{"annotated writable", writeTool("create_dashboard"), false},
		{"nil annotations", unannotatedTool("some_tool"), false},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := IsReadOnlyTool(tt.tool); got != tt.want {
				t.Errorf("IsReadOnlyTool(%s) = %v, want %v", tt.tool.Name, got, tt.want)
			}
		})
	}
}

func TestCanAccessTool(t *testing.T) {
	ro := readOnlyTool("get_dashboard")
	wr := writeTool("create_dashboard")
	un := unannotatedTool("unknown_tool")

	tests := []struct {
		name string
		role string
		tool mcp.Tool
		want bool
	}{
		{"admin read-only", "Admin", ro, true},
		{"admin writable", "Admin", wr, true},
		{"admin unannotated", "Admin", un, true},
		{"editor read-only", "Editor", ro, true},
		{"editor writable", "Editor", wr, true},
		{"editor unannotated", "Editor", un, true},
		{"viewer read-only", "Viewer", ro, true},
		{"viewer writable", "Viewer", wr, false},
		{"viewer unannotated", "Viewer", un, false},
		{"unknown role writable", "Unknown", wr, false},
		{"unknown role read-only", "Unknown", ro, true},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := CanAccessTool(tt.role, tt.tool); got != tt.want {
				t.Errorf("CanAccessTool(%s, %s) = %v, want %v", tt.role, tt.tool.Name, got, tt.want)
			}
		})
	}
}

func TestFilterToolsByRole(t *testing.T) {
	tools := []mcp.Tool{
		readOnlyTool("get_dashboard"),
		writeTool("create_dashboard"),
		readOnlyTool("list_datasources"),
		writeTool("delete_dashboard"),
		unannotatedTool("unknown_tool"),
	}

	t.Run("admin gets all", func(t *testing.T) {
		filtered := FilterToolsByRole(tools, "Admin")
		if len(filtered) != 5 {
			t.Errorf("Admin got %d tools, want 5", len(filtered))
		}
	})

	t.Run("editor gets all", func(t *testing.T) {
		filtered := FilterToolsByRole(tools, "Editor")
		if len(filtered) != 5 {
			t.Errorf("Editor got %d tools, want 5", len(filtered))
		}
	})

	t.Run("viewer gets read-only only", func(t *testing.T) {
		filtered := FilterToolsByRole(tools, "Viewer")
		if len(filtered) != 2 {
			t.Errorf("Viewer got %d tools, want 2", len(filtered))
		}
		for _, tool := range filtered {
			if !IsReadOnlyTool(tool) {
				t.Errorf("Viewer got non-read-only tool: %s", tool.Name)
			}
		}
	})

	t.Run("empty list", func(t *testing.T) {
		filtered := FilterToolsByRole([]mcp.Tool{}, "Viewer")
		if len(filtered) != 0 {
			t.Errorf("Expected empty, got %d", len(filtered))
		}
	})
}

File: ask-o11y/src/utils/__tests__/rbac.test.ts
import { isReadOnlyTool, canAccessTool, filterToolsByRole } from '../rbac';

const readOnly = (name: string) => ({ name, annotations: { readOnlyHint: true } });
const writable = (name: string) => ({ name, annotations: { readOnlyHint: false } });
const unannotated = (name: string) => ({ name });

describe('RBAC Utilities', () => {
  describe('isReadOnlyTool', () => {
    it('should return true for tools annotated as read-only', () => {
      expect(isReadOnlyTool(readOnly('get_dashboard'))).toBe(true);
      expect(isReadOnlyTool(readOnly('list_datasources'))).toBe(true);
    });

    it('should return false for tools annotated as writable', () => {
      expect(isReadOnlyTool(writable('create_dashboard'))).toBe(false);
      expect(isReadOnlyTool(writable('delete_dashboard'))).toBe(false);
    });

    it('should return false for tools without annotations', () => {
      expect(isReadOnlyTool(unannotated('some_tool'))).toBe(false);
    });
  });

  describe('canAccessTool', () => {
    it('should allow Admin access to all tools', () => {
      expect(canAccessTool('Admin', readOnly('get_dashboard'))).toBe(true);
      expect(canAccessTool('Admin', writable('create_dashboard'))).toBe(true);
      expect(canAccessTool('Admin', unannotated('some_tool'))).toBe(true);
    });

    it('should allow Editor access to all tools', () => {
      expect(canAccessTool('Editor', readOnly('get_dashboard'))).toBe(true);
      expect(canAccessTool('Editor', writable('create_dashboard'))).toBe(true);
      expect(canAccessTool('Editor', unannotated('some_tool'))).toBe(true);
    });

    it('should allow Viewer access to read-only tools only', () => {
      expect(canAccessTool('Viewer', readOnly('get_dashboard'))).toBe(true);
      expect(canAccessTool('Viewer', readOnly('list_datasources'))).toBe(true);
    });

    it('should deny Viewer access to writable tools', () => {
      expect(canAccessTool('Viewer', writable('create_dashboard'))).toBe(false);
      expect(canAccessTool('Viewer', writable('delete_dashboard'))).toBe(false);
    });

    it('should deny Viewer access to unannotated tools', () => {
      expect(canAccessTool('Viewer', unannotated('some_tool'))).toBe(false);
    });

    it('should treat unknown roles like Viewer', () => {
      expect(canAccessTool('Unknown', readOnly('get_dashboard'))).toBe(true);
      expect(canAccessTool('Unknown', writable('create_dashboard'))).toBe(false);
    });
  });

  describe('filterToolsByRole', () => {
    const mockTools = [
      readOnly('get_dashboard'),
      writable('create_dashboard'),
      readOnly('list_datasources'),
      writable('update_datasource'),
      unannotated('external_tool'),
    ];

    it('should return all tools for Admin', () => {
      const filtered = filterToolsByRole(mockTools, 'Admin');
      expect(filtered).toHaveLength(5);
      expect(filtered).toEqual(mockTools);
    });

    it('should return all tools for Editor', () => {
      const filtered = filterToolsByRole(mockTools, 'Editor');
      expect(filtered).toHaveLength(5);
      expect(filtered).toEqual(mockTools);
    });

    it('should return only read-only tools for Viewer', () => {
      const filtered = filterToolsByRole(mockTools, 'Viewer');
      expect(filtered).toHaveLength(2);
      expect(filtered.map((t) => t.name)).toEqual(['get_dashboard', 'list_datasources']);
    });

    it('should handle empty tool list', () => {
      expect(filterToolsByRole([], 'Viewer')).toEqual([]);
      expect(filterToolsByRole([], 'Admin')).toEqual([]);
    });

    it('should preserve additional properties on tools', () => {
      const toolsWithMetadata = [
        { ...readOnly('get_dashboard'), description: 'Get dashboard', version: '1.0' },
        { ...writable('create_dashboard'), description: 'Create dashboard', version: '1.0' },
      ];

      const filtered = filterToolsByRole(toolsWithMetadata, 'Viewer');
      expect(filtered).toHaveLength(1);
      expect(filtered[0]).toEqual({
        name: 'get_dashboard',
        annotations: { readOnlyHint: true },
        description: 'Get dashboard',
        version: '1.0',
      });
    });

    it('should treat unknown roles like Viewer', () => {
      const filtered = filterToolsByRole(mockTools, 'UnknownRole');
      expect(filtered).toHaveLength(2);
    });
  });
});
